dali/kernels/signal/fft/stft_gpu_impl.cu

```diff
@@ -0,0 +1,299 @@
+  if (args != args_) {
+    Reset();
+    args_ = args;
+  }
+
+  int64_t windows = 0;
+  TensorListShape<> shape;
+  shape.resize(lengths.size(), 2);
+
+  transform_in_.data.clear();
+  transform_out_.data.clear();
+  transform_in_.resize(lengths.size(), 2);
+  transform_out_.resize(lengths.size(), 2);
+  int N = lengths.size();
+  for (int i = 0; i < N; i++) {
+    int64_t l = lengths[i];
+    int64_t n = args_.num_windows(l);
+    TensorShape<2> ts_in = { n, transform_in_size() };
+    TensorShape<2> ts_out = { n, transform_out_size() };
+    transform_out_.shape.set_tensor_shape(i, ts_out);
+    transform_in_.shape.set_tensor_shape(i, ts_in);
+
+    if (!args_.time_major_layout)
+      std::swap(ts_out[0], ts_out[1]);
+    shape.set_tensor_shape(i, ts_out);
+    windows += n;
+  }
+  total_windows_ = windows;
+
+  KernelRequirements req;
+  ScratchpadEstimator se;
+
+  SetupWindowExtraction(ctx, se, lengths);
+  CreatePlans(windows);
+  ReserveTempStorage(se);
+  SetupPostprocessing(ctx, se);
+
+  req.output_shapes = { shape };
+  req.scratch_sizes = se.sizes;
+  return req;
+}
+
+void StftImplGPU::SetupWindowExtraction(
+    KernelContext &ctx,
+    ScratchpadEstimator &se,
+    span<const int64_t> input_lengths) {
+  ExtractWindowsBatchedArgs extract_args;
+  static_cast<ExtractWindowsArgs &>(extract_args) = static_cast<ExtractWindowsArgs &>(args_);
+  extract_args.output_window_length = transform_in_size();  // include padding, if necessary
+
+  KernelRequirements extract_req = window_extractor_.Setup(ctx, input_lengths, extract_args);
+  add_scratch(se, extract_req.scratch_sizes);
+}
+
+void StftImplGPU::SetupPostprocessing(KernelContext &ctx, ScratchpadEstimator &se) {
+  if (args_.spectrum_type == FFT_SPECTRUM_COMPLEX) {
+    post_real_.reset();
+    post_complex_ = fft_postprocess::GetSTFTPostprocessor(args_.time_major_layout);
+    auto req = post_complex_->Setup(ctx, transform_out_.shape);
+    add_scratch(se, req.scratch_sizes);
+  } else {
+    post_complex_.reset();
+    post_real_ = fft_postprocess::GetSpectrogramPostprocessor(
+        args_.time_major_layout, args_.spectrum_type);
+    auto req = post_real_->Setup(ctx, transform_out_.shape);
+    add_scratch(se, req.scratch_sizes);
+  }
+}
+
+
+void StftImplGPU::CreatePlans(int64_t nwindows) {
+  int64_t max_windows = kMaxSize;
+
+  while (max_windows * transform_size() > kMaxSize)
+    max_windows >>= 1;
+
+  while (max_windows > nwindows)
+    max_windows >>= 1;
+
+  if (max_windows == 0)
+    max_windows = 1;
+
+  max_windows_ = max_windows;
+  min_windows_ = std::min(max_windows_, next_pow2(kMinSize / transform_size()));
+
+  int n[1] = { transform_size() };
+  for (int w = max_windows_; w >= min_windows_; w >>= 1) {
+    auto &plan = plans_[w];
+    if (!plan.handle) {
+      cufftHandle handle;
+      CUDA_CALL(cufftCreate(&handle));
+      plan.handle.reset(handle);
+      CUDA_CALL(cufftSetAutoAllocation(handle, 0));
```

Comment:
0 or false?

Reply:
The signature says "int", but it's C. I can make it false.

Reply:
Just a question.